<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.157">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ralf Gabriels">
<meta name="dcterms.date" content="2023-02-16">

<title>ProteomicsML - NIST (part 2): Traditional ML: Gradient boosting</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../static/img/proteomicsml-icon.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-VVRMZHT2W2"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
 
  gtag('consent', 'default', {
    'ad_storage': 'denied',
    'analytics_storage': 'denied'
  });
gtag('config', 'G-VVRMZHT2W2', { 'anonymize_ip': true});
</script>
<style>html{ scroll-behavior: smooth; }</style>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../static/css/custom-styles.css">
<meta property="og:title" content="ProteomicsML - NIST (part 2): Traditional ML: Gradient boosting">
<meta property="og:description" content="This is the second part in a three-part tutorial. You are recommended to start with the first section, where the NIST spectral library is parsed and prepared for use in the second and third parts.">
<meta property="og:site-name" content="ProteomicsML">
<meta name="twitter:title" content="ProteomicsML - NIST (part 2): Traditional ML: Gradient boosting">
<meta name="twitter:description" content="This is the second part in a three-part tutorial. You are recommended to start with the first section, where the NIST spectral library is parsed and prepared for use in the second and third parts.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../static/img/proteomicsml-logo-inverse.svg" alt="ProteomicsML logo" class="navbar-logo">
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../index.html" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../publication.html">
 <span class="menu-text">Publication</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../tutorials/index.html">
 <span class="menu-text">Tutorials</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../datasets/index.html">
 <span class="menu-text">Datasets</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../contributing.html">
 <span class="menu-text">Contributing</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../code-of-conduct.html">
 <span class="menu-text">Code of Conduct</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ProteomicsML"><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/ProteomicsML"><i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">NIST (part 2): Traditional ML: Gradient boosting</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">Home</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../publication.html" class="sidebar-item-text sidebar-link">Publication</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../tutorials/index.html" class="sidebar-item-text sidebar-link">Tutorials</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">Detectability</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/detectability/modeling-protein-detectability.html" class="sidebar-item-text sidebar-link">Modelling protein detectability with an MLP</a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Fragmentation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/nist-1-parsing-spectral-library.html" class="sidebar-item-text sidebar-link">NIST (part 1): Preparing a spectral library for ML</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.html" class="sidebar-item-text sidebar-link active">NIST (part 2): Traditional ML: Gradient boosting</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/nist-3-deep-learning-lstm.html" class="sidebar-item-text sidebar-link">NIST (part 3): Deep learning: BiLSTM</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/preannotated-prosit.html" class="sidebar-item-text sidebar-link">Prosit-style GRU with pre-annotated ProteomeTools data</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/raw-to-prosit.html" class="sidebar-item-text sidebar-link">Raw file processing with PROSIT style annotation</a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">Ion mobility</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/ionmobility/meier-tims-ccs.html" class="sidebar-item-text sidebar-link">Predicting CCS values for TIMS data</a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">Retention time</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/dlomix-prosit-rt.html" class="sidebar-item-text sidebar-link">DLOmix embedding of Prosit model on ProteomeTools data</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/manual-prosit-rt.html" class="sidebar-item-text sidebar-link">Manual embedding of Bi-LSTM model on ProteomeTools data</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/mq-evidence-to-ml.html" class="sidebar-item-text sidebar-link">Preparing a retention time data set for machine learning</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/deeplc-transfer-learning.html" class="sidebar-item-text sidebar-link">Transfer learning with DeepLC</a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../datasets/index.html" class="sidebar-item-text sidebar-link">Datasets</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false">Detectability</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/detectability/ArabidopsisLightDarkProteome.html" class="sidebar-item-text sidebar-link">Arabidopsis PeptideAtlas Light and Dark Proteome</a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="false">Fragmentation</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/fragmentation/nist.html" class="sidebar-item-text sidebar-link">NIST Peptide libraries</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/fragmentation/ProteomeTools_FI.html" class="sidebar-item-text sidebar-link">ProteomeTools</a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="false">Ion mobility</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/ionmobility/Meier_TIMS.html" class="sidebar-item-text sidebar-link">Meier et al.&nbsp;TIMS</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/ionmobility/VanPuyvelde_TWIMS.html" class="sidebar-item-text sidebar-link">Van Puyvelde et al.&nbsp;TWIMS</a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" aria-expanded="false">Retention time</a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" aria-expanded="false">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-10" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/retentiontime/DLOmix_RT.html" class="sidebar-item-text sidebar-link">DLOmix</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/retentiontime/ProteomeTools_RT.html" class="sidebar-item-text sidebar-link">ProteomeTools</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/retentiontime/PXD028248_RT.html" class="sidebar-item-text sidebar-link">PXD028248</a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../contributing.html" class="sidebar-item-text sidebar-link">Contributing</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../code-of-conduct.html" class="sidebar-item-text sidebar-link">Code of Conduct</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">1. Introduction</a></li>
  <li><a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation">2 Data preparation</a>
  <ul>
  <li><a href="#feature-engineering" id="toc-feature-engineering" class="nav-link" data-scroll-target="#feature-engineering">2.1 Feature engineering</a></li>
  <li><a href="#getting-the-target-intensities" id="toc-getting-the-target-intensities" class="nav-link" data-scroll-target="#getting-the-target-intensities">2.2 Getting the target intensities</a></li>
  <li><a href="#bringing-it-all-together" id="toc-bringing-it-all-together" class="nav-link" data-scroll-target="#bringing-it-all-together">2.3 Bringing it all together</a></li>
  </ul></li>
  <li><a href="#training-the-model" id="toc-training-the-model" class="nav-link" data-scroll-target="#training-the-model">3 Training the model</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ProteomicsML/ProteomicsML/blob/main/tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.ipynb" class="toc-action">View source</a></p><p><a href="https://github.dev/ProteomicsML/ProteomicsML/blob/main/tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.ipynb" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">NIST (part 2): Traditional ML: Gradient boosting</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading"></div>
  
    <div class="quarto-title-meta-contents">
    Ralf Gabriels <a href="https://orcid.org/0000-0002-1679-1711" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            VIB-UGent Center for Medical Biotechnology, VIB, Belgium
          </p>
        <p class="affiliation">
            Department of Biomolecular Medicine, Ghent University, Belgium
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 16, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://colab.research.google.com/github/ProteomicsML/ProteomicsML/blob/main/tutorials/fragmentation/_nist-2-traditional-ml-gradient-boosting.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">1. Introduction</h2>
<p>This is the second part in a three-part tutorial. You are recommended to start with the first section, where the NIST spectral library is parsed and prepared for use in the second and third parts.</p>
<ol type="1">
<li><a href="https://www.proteomicsml.org/tutorials/fragmentation/nist-1-parsing-spectral-library.html">Preparing a spectral library for ML</a></li>
<li><a href="https://www.proteomicsml.org/tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.html">Traditional ML: Gradient boosting</a></li>
<li><a href="https://www.proteomicsml.org/tutorials/fragmentation/nist-3-deep-learning-lstm.html">Deep learning: BiLSTM</a></li>
</ol>
<p>In this tutorial, you will learn how to build a fragmentation intensity predictor similar to MS²PIP v3 <span class="citation" data-cites="Gabriels2019">(<a href="#ref-Gabriels2019" role="doc-biblioref">Gabriels, Martens, and Degroeve 2019</a>)</span> with traditional machine learning (ML) feature engineering and Gradient Boosting <span class="citation" data-cites="Friedman2002">(<a href="#ref-Friedman2002" role="doc-biblioref">Friedman 2002</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Installing required python packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> pip install rich<span class="op">~=</span><span class="fl">12.5</span> numpy<span class="op">~=</span><span class="fl">1.21</span> pandas<span class="op">~=</span><span class="fl">1.3</span> matplotlib<span class="op">~=</span><span class="fl">3.5</span> seaborn<span class="op">~=</span><span class="fl">0.11</span> scikit<span class="op">-</span>learn<span class="op">~=</span><span class="fl">1.0</span> <span class="op">--</span>quiet</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-preparation" class="level2">
<h2 class="anchored" data-anchor-id="data-preparation">2 Data preparation</h2>
<p>We will use the spectral library that was already parsed in part 1 of this tutorial series.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>train_val_spectra <span class="op">=</span> pd.read_feather(<span class="st">"ftp://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-trainval.feather"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>test_spectra <span class="op">=</span> pd.read_feather(<span class="st">"ftp://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-test.feather"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="feature-engineering" class="level3">
<h3 class="anchored" data-anchor-id="feature-engineering">2.1 Feature engineering</h3>
<p>In traditional ML, the input features for the algorithm usually require some engineering. For fragmentation intensity prediction this is not different. Following the MS²PIP methods, we will calculate the distributions of several amino acid properties across the peptide and fragment ion sequences.</p>
<p>Using the distribution of these properties instead of the actual properties per amino acid allows MS²PIP to get a fixed length feature matrix for input peptides with varying lengths.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> rich <span class="im">import</span> progress</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>amino_acids <span class="op">=</span> <span class="bu">list</span>(<span class="st">"ACDEFGHIKLMNPQRSTVWY"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>aa_properties <span class="op">=</span> {</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"basicity"</span>: np.array([<span class="dv">37</span>,<span class="dv">35</span>,<span class="dv">59</span>,<span class="dv">129</span>,<span class="dv">94</span>,<span class="dv">0</span>,<span class="dv">210</span>,<span class="dv">81</span>,<span class="dv">191</span>,<span class="dv">81</span>,<span class="dv">106</span>,<span class="dv">101</span>,<span class="dv">117</span>,<span class="dv">115</span>,<span class="dv">343</span>,<span class="dv">49</span>,<span class="dv">90</span>,<span class="dv">60</span>,<span class="dv">134</span>,<span class="dv">104</span>]),</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"helicity"</span>: np.array([<span class="dv">68</span>,<span class="dv">23</span>,<span class="dv">33</span>,<span class="dv">29</span>,<span class="dv">70</span>,<span class="dv">58</span>,<span class="dv">41</span>,<span class="dv">73</span>,<span class="dv">32</span>,<span class="dv">73</span>,<span class="dv">66</span>,<span class="dv">38</span>,<span class="dv">0</span>,<span class="dv">40</span>,<span class="dv">39</span>,<span class="dv">44</span>,<span class="dv">53</span>,<span class="dv">71</span>,<span class="dv">51</span>,<span class="dv">55</span>]),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"hydrophobicity"</span>: np.array([<span class="dv">51</span>,<span class="dv">75</span>,<span class="dv">25</span>,<span class="dv">35</span>,<span class="dv">100</span>,<span class="dv">16</span>,<span class="dv">3</span>,<span class="dv">94</span>,<span class="dv">0</span>,<span class="dv">94</span>,<span class="dv">82</span>,<span class="dv">12</span>,<span class="dv">0</span>,<span class="dv">22</span>,<span class="dv">22</span>,<span class="dv">21</span>,<span class="dv">39</span>,<span class="dv">80</span>,<span class="dv">98</span>,<span class="dv">70</span>]),</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"pI"</span>: np.array([<span class="dv">32</span>,<span class="dv">23</span>,<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">27</span>,<span class="dv">32</span>,<span class="dv">48</span>,<span class="dv">32</span>,<span class="dv">69</span>,<span class="dv">32</span>,<span class="dv">29</span>,<span class="dv">26</span>,<span class="dv">35</span>,<span class="dv">28</span>,<span class="dv">79</span>,<span class="dv">29</span>,<span class="dv">28</span>,<span class="dv">31</span>,<span class="dv">31</span>,<span class="dv">28</span>]),</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>properties_df <span class="op">=</span> pd.DataFrame(aa_properties, index<span class="op">=</span>amino_acids)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>properties_df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>basicity</th>
      <th>helicity</th>
      <th>hydrophobicity</th>
      <th>pI</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A</th>
      <td>37</td>
      <td>68</td>
      <td>51</td>
      <td>32</td>
    </tr>
    <tr>
      <th>C</th>
      <td>35</td>
      <td>23</td>
      <td>75</td>
      <td>23</td>
    </tr>
    <tr>
      <th>D</th>
      <td>59</td>
      <td>33</td>
      <td>25</td>
      <td>0</td>
    </tr>
    <tr>
      <th>E</th>
      <td>129</td>
      <td>29</td>
      <td>35</td>
      <td>4</td>
    </tr>
    <tr>
      <th>F</th>
      <td>94</td>
      <td>70</td>
      <td>100</td>
      <td>27</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>properties <span class="op">=</span> np.array([</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">37</span>,<span class="dv">35</span>,<span class="dv">59</span>,<span class="dv">129</span>,<span class="dv">94</span>,<span class="dv">0</span>,<span class="dv">210</span>,<span class="dv">81</span>,<span class="dv">191</span>,<span class="dv">81</span>,<span class="dv">106</span>,<span class="dv">101</span>,<span class="dv">117</span>,<span class="dv">115</span>,<span class="dv">343</span>,<span class="dv">49</span>,<span class="dv">90</span>,<span class="dv">60</span>,<span class="dv">134</span>,<span class="dv">104</span>],  <span class="co"># basicity</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">68</span>,<span class="dv">23</span>,<span class="dv">33</span>,<span class="dv">29</span>,<span class="dv">70</span>,<span class="dv">58</span>,<span class="dv">41</span>,<span class="dv">73</span>,<span class="dv">32</span>,<span class="dv">73</span>,<span class="dv">66</span>,<span class="dv">38</span>,<span class="dv">0</span>,<span class="dv">40</span>,<span class="dv">39</span>,<span class="dv">44</span>,<span class="dv">53</span>,<span class="dv">71</span>,<span class="dv">51</span>,<span class="dv">55</span>],  <span class="co"># helicity</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">51</span>,<span class="dv">75</span>,<span class="dv">25</span>,<span class="dv">35</span>,<span class="dv">100</span>,<span class="dv">16</span>,<span class="dv">3</span>,<span class="dv">94</span>,<span class="dv">0</span>,<span class="dv">94</span>,<span class="dv">82</span>,<span class="dv">12</span>,<span class="dv">0</span>,<span class="dv">22</span>,<span class="dv">22</span>,<span class="dv">21</span>,<span class="dv">39</span>,<span class="dv">80</span>,<span class="dv">98</span>,<span class="dv">70</span>],  <span class="co"># hydrophobicity</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">32</span>,<span class="dv">23</span>,<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">27</span>,<span class="dv">32</span>,<span class="dv">48</span>,<span class="dv">32</span>,<span class="dv">69</span>,<span class="dv">32</span>,<span class="dv">29</span>,<span class="dv">26</span>,<span class="dv">35</span>,<span class="dv">28</span>,<span class="dv">79</span>,<span class="dv">29</span>,<span class="dv">28</span>,<span class="dv">31</span>,<span class="dv">31</span>,<span class="dv">28</span>],  <span class="co"># pI</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>])</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>quantiles <span class="op">=</span> [<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="dv">1</span>]</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>aa_indices <span class="op">=</span> {aa: i <span class="cf">for</span> i, aa <span class="kw">in</span>  <span class="bu">enumerate</span>(<span class="st">"ACDEFGHIKLMNPQRSTVWY"</span>)}</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>aa_to_index <span class="op">=</span> np.vectorize(<span class="kw">lambda</span> aa: aa_indices[aa])</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> encode_peptide(sequence, charge):</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 4 properties * 5 quantiles * 3 ion types + 4 properties * 4 site + 2 global</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    n_features <span class="op">=</span> <span class="dv">78</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>    n_ions <span class="op">=</span> <span class="bu">len</span>(sequence) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Encode amino acids as integers to index amino acid properties for peptide sequence</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    peptide_indexed <span class="op">=</span> aa_to_index(np.array(<span class="bu">list</span>(sequence)))</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    peptide_properties <span class="op">=</span> properties[:, peptide_indexed]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Empty peptide_features array</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>    peptide_features <span class="op">=</span> np.full((n_ions, n_features), np.nan)</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> b_ion_number <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, n_ions <span class="op">+</span> <span class="dv">1</span>):</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate quantiles of features across peptide, b-ion, and y-ion</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>        peptide_quantiles <span class="op">=</span> np.hstack(</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>            np.quantile(peptide_properties, quantiles, axis<span class="op">=</span><span class="dv">1</span>).transpose()</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>        b_ion_quantiles <span class="op">=</span> np.hstack(</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>            np.quantile(peptide_properties[:,:b_ion_number], quantiles, axis<span class="op">=</span><span class="dv">1</span>).transpose()</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>        y_ion_quantiles <span class="op">=</span> np.hstack(</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>            np.quantile(peptide_properties[:,b_ion_number:], quantiles, axis<span class="op">=</span><span class="dv">1</span>).transpose()</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Properties on specific sites: nterm, frag-1, frag+1, cterm</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>        specific_site_indexes <span class="op">=</span> np.array([<span class="dv">0</span>, b_ion_number <span class="op">-</span> <span class="dv">1</span>, b_ion_number, <span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>        specific_site_properties <span class="op">=</span> np.hstack(peptide_properties[:, specific_site_indexes].transpose())</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Global features: Length and charge</span></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>        global_features <span class="op">=</span> np.array([<span class="bu">len</span>(sequence), <span class="bu">int</span>(charge)])</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Assign to peptide_features array</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">0</span>:<span class="dv">20</span>] <span class="op">=</span> peptide_quantiles</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">20</span>:<span class="dv">40</span>] <span class="op">=</span> b_ion_quantiles</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">40</span>:<span class="dv">60</span>] <span class="op">=</span> y_ion_quantiles</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">60</span>:<span class="dv">76</span>] <span class="op">=</span> specific_site_properties</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">76</span>:<span class="dv">78</span>] <span class="op">=</span> global_features</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> peptide_features</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_feature_names():</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a>    feature_names <span class="op">=</span> []</span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> level <span class="kw">in</span> [<span class="st">"peptide"</span>, <span class="st">"b"</span>, <span class="st">"y"</span>]:</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> aa_property <span class="kw">in</span> [<span class="st">"basicity"</span>, <span class="st">"helicity"</span>, <span class="st">"hydrophobicity"</span>, <span class="st">"pi"</span>]:</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> quantile <span class="kw">in</span> [<span class="st">"min"</span>, <span class="st">"q1"</span>, <span class="st">"q2"</span>, <span class="st">"q3"</span>, <span class="st">"max"</span>]:</span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>                feature_names.append(<span class="st">"_"</span>.join([level, aa_property, quantile]))</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> site <span class="kw">in</span> [<span class="st">"nterm"</span>, <span class="st">"fragmin1"</span>, <span class="st">"fragplus1"</span>, <span class="st">"cterm"</span>]:</span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> aa_property <span class="kw">in</span> [<span class="st">"basicity"</span>, <span class="st">"helicity"</span>, <span class="st">"hydrophobicity"</span>, <span class="st">"pi"</span>]:</span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>            feature_names.append(<span class="st">"_"</span>.join([site, aa_property]))</span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>    feature_names.extend([<span class="st">"length"</span>, <span class="st">"charge"</span>])</span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> feature_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s test it with a single peptide. Feel free to use your own name as a “peptide”; as long as it does not contain any non-amino acid characters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>peptide_features <span class="op">=</span> pd.DataFrame(encode_peptide(<span class="st">"RALFGARIELS"</span>, <span class="dv">2</span>), columns<span class="op">=</span>generate_feature_names())</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>peptide_features</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>peptide_basicity_min</th>
      <th>peptide_basicity_q1</th>
      <th>peptide_basicity_q2</th>
      <th>peptide_basicity_q3</th>
      <th>peptide_basicity_max</th>
      <th>peptide_helicity_min</th>
      <th>peptide_helicity_q1</th>
      <th>peptide_helicity_q2</th>
      <th>peptide_helicity_q3</th>
      <th>peptide_helicity_max</th>
      <th>...</th>
      <th>fragplus1_basicity</th>
      <th>fragplus1_helicity</th>
      <th>fragplus1_hydrophobicity</th>
      <th>fragplus1_pi</th>
      <th>cterm_basicity</th>
      <th>cterm_helicity</th>
      <th>cterm_hydrophobicity</th>
      <th>cterm_pi</th>
      <th>length</th>
      <th>charge</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>37.0</td>
      <td>68.0</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>81.0</td>
      <td>73.0</td>
      <td>94.0</td>
      <td>32.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>94.0</td>
      <td>70.0</td>
      <td>100.0</td>
      <td>27.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>58.0</td>
      <td>16.0</td>
      <td>32.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>37.0</td>
      <td>68.0</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>81.0</td>
      <td>73.0</td>
      <td>94.0</td>
      <td>32.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>129.0</td>
      <td>29.0</td>
      <td>35.0</td>
      <td>4.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>81.0</td>
      <td>73.0</td>
      <td>94.0</td>
      <td>32.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.0</td>
      <td>43.0</td>
      <td>81.0</td>
      <td>111.5</td>
      <td>343.0</td>
      <td>29.0</td>
      <td>41.5</td>
      <td>68.0</td>
      <td>71.5</td>
      <td>73.0</td>
      <td>...</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>49.0</td>
      <td>44.0</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>11.0</td>
      <td>2.0</td>
    </tr>
  </tbody>
</table>
<p>10 rows × 78 columns</p>
</div>
</div>
</div>
</section>
<section id="getting-the-target-intensities" class="level3">
<h3 class="anchored" data-anchor-id="getting-the-target-intensities">2.2 Getting the target intensities</h3>
<p>The target intensities are the observed intensities which the model will learn to predict. Let’s first try with a single spectrum.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>test_spectrum <span class="op">=</span> train_val_spectra.iloc[<span class="dv">4</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>peptide_targets <span class="op">=</span>  pd.DataFrame({</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"b_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>],</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"y_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>],</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>peptide_targets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>b_target</th>
      <th>y_target</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.000000</td>
      <td>0.118507</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.229717</td>
      <td>0.079770</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.294631</td>
      <td>0.088712</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.234662</td>
      <td>0.145900</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.185732</td>
      <td>0.205005</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.134395</td>
      <td>0.261630</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.081856</td>
      <td>0.305119</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.043793</td>
      <td>0.296351</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.000000</td>
      <td>0.205703</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.000000</td>
      <td>0.155991</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<p>These are the intensities for the b- and y-ions, each ordered from 1 to 9. In MS²PIP, however, a clever trick is applied to reuse the computed features for each <em>fragment ion pair</em>. Doing so makes perfect sense, as both ions in such a fragment ion pair originated from the same fragmentation event. For this peptide, the fragment ion pairs are b1-y9, b2-y8, b3-y7, etc. To match all of the pairs, we simply have to reverse the y-ion series intensities:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>peptide_targets <span class="op">=</span>  pd.DataFrame({</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"b_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>],</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"y_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>][::<span class="op">-</span><span class="dv">1</span>],</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>peptide_targets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>b_target</th>
      <th>y_target</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.229717</td>
      <td>0.155991</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.294631</td>
      <td>0.205703</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.234662</td>
      <td>0.296351</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.185732</td>
      <td>0.305119</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.134395</td>
      <td>0.261630</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.081856</td>
      <td>0.205005</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.043793</td>
      <td>0.145900</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.000000</td>
      <td>0.088712</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.000000</td>
      <td>0.079770</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0.000000</td>
      <td>0.118507</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
</section>
<section id="bringing-it-all-together" class="level3">
<h3 class="anchored" data-anchor-id="bringing-it-all-together">2.3 Bringing it all together</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>features <span class="op">=</span> encode_peptide(test_spectrum[<span class="st">"sequence"</span>], test_spectrum[<span class="st">"charge"</span>])</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>targets <span class="op">=</span> np.stack([test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>], test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>][::<span class="op">-</span><span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>spectrum_id <span class="op">=</span> np.full(shape<span class="op">=</span>(targets.shape[<span class="dv">0</span>], <span class="dv">1</span>), fill_value<span class="op">=</span>test_spectrum[<span class="st">"index"</span>])  <span class="co"># Repeat id for all ions</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(np.hstack([spectrum_id, features, targets]), columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>] <span class="op">+</span> generate_feature_names() <span class="op">+</span> [<span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>spectrum_id</th>
      <th>peptide_basicity_min</th>
      <th>peptide_basicity_q1</th>
      <th>peptide_basicity_q2</th>
      <th>peptide_basicity_q3</th>
      <th>peptide_basicity_max</th>
      <th>peptide_helicity_min</th>
      <th>peptide_helicity_q1</th>
      <th>peptide_helicity_q2</th>
      <th>peptide_helicity_q3</th>
      <th>...</th>
      <th>fragplus1_hydrophobicity</th>
      <th>fragplus1_pi</th>
      <th>cterm_basicity</th>
      <th>cterm_helicity</th>
      <th>cterm_hydrophobicity</th>
      <th>cterm_pi</th>
      <th>length</th>
      <th>charge</th>
      <th>b_target</th>
      <th>y_target</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.229717</td>
      <td>0.155991</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.294631</td>
      <td>0.205703</td>
    </tr>
    <tr>
      <th>3</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.234662</td>
      <td>0.296351</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.185732</td>
      <td>0.305119</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.134395</td>
      <td>0.261630</td>
    </tr>
    <tr>
      <th>6</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.081856</td>
      <td>0.205005</td>
    </tr>
    <tr>
      <th>7</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.043793</td>
      <td>0.145900</td>
    </tr>
    <tr>
      <th>8</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>80.0</td>
      <td>31.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.000000</td>
      <td>0.088712</td>
    </tr>
    <tr>
      <th>9</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>21.0</td>
      <td>29.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.000000</td>
      <td>0.079770</td>
    </tr>
    <tr>
      <th>10</th>
      <td>5.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>40.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>12.0</td>
      <td>2.0</td>
      <td>0.000000</td>
      <td>0.118507</td>
    </tr>
  </tbody>
</table>
<p>11 rows × 81 columns</p>
</div>
</div>
</div>
<p>The following function applies these steps over a collection of spectra and returns the full feature/target table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_ml_input(spectra):</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    tables <span class="op">=</span> []</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> spectrum <span class="kw">in</span> progress.track(spectra.to_dict(orient<span class="op">=</span><span class="st">"records"</span>)):</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>        features <span class="op">=</span> encode_peptide(spectrum[<span class="st">"sequence"</span>], spectrum[<span class="st">"charge"</span>])</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>        targets <span class="op">=</span> np.stack([spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>], spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>][::<span class="op">-</span><span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>        spectrum_id <span class="op">=</span> np.full(shape<span class="op">=</span>(targets.shape[<span class="dv">0</span>], <span class="dv">1</span>), fill_value<span class="op">=</span>spectrum[<span class="st">"index"</span>])  <span class="co"># Repeat id for all ions</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>        table <span class="op">=</span> np.hstack([spectrum_id, features, targets])</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>        tables.append(table)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    full_table <span class="op">=</span> np.vstack(tables)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    spectra_encoded <span class="op">=</span> pd.DataFrame(full_table, columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>] <span class="op">+</span> generate_feature_names() <span class="op">+</span> [<span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> spectra_encoded</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this might take some time, sometimes up to 30 minutes. To skip this step, simple download the file with pre-encoded features and targets, and load in two cells below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>train_val_encoded <span class="op">=</span> generate_ml_input(train_val_spectra)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>train_val_encoded.to_feather(<span class="st">"fragmentation-nist-humanhcd20160503-parsed-trainval-encoded.feather"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>test_encoded <span class="op">=</span> generate_ml_input(test_spectra)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>test_encoded.to_feather(<span class="st">"fragmentation-nist-humanhcd20160503-parsed-test-encoded.feather"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uncomment this step to load pre-encoded features from a file:</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">#train_val_encoded = pd.read_feather("ftp://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-trainval-encoded.feather")</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#test_encoded = pd.read_feather("ftp://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-test-encoded.feather")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>train_val_encoded</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>spectrum_id</th>
      <th>peptide_basicity_min</th>
      <th>peptide_basicity_q1</th>
      <th>peptide_basicity_q2</th>
      <th>peptide_basicity_q3</th>
      <th>peptide_basicity_max</th>
      <th>peptide_helicity_min</th>
      <th>peptide_helicity_q1</th>
      <th>peptide_helicity_q2</th>
      <th>peptide_helicity_q3</th>
      <th>...</th>
      <th>fragplus1_hydrophobicity</th>
      <th>fragplus1_pi</th>
      <th>cterm_basicity</th>
      <th>cterm_helicity</th>
      <th>cterm_hydrophobicity</th>
      <th>cterm_pi</th>
      <th>length</th>
      <th>charge</th>
      <th>b_target</th>
      <th>y_target</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>0.0</td>
      <td>69.0</td>
      <td>22.0</td>
      <td>2.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>0.0</td>
      <td>69.0</td>
      <td>22.0</td>
      <td>2.0</td>
      <td>0.094060</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>0.0</td>
      <td>69.0</td>
      <td>22.0</td>
      <td>2.0</td>
      <td>0.180642</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>0.0</td>
      <td>69.0</td>
      <td>22.0</td>
      <td>2.0</td>
      <td>0.204203</td>
      <td>0.050476</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>37.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>68.0</td>
      <td>...</td>
      <td>51.0</td>
      <td>32.0</td>
      <td>191.0</td>
      <td>32.0</td>
      <td>0.0</td>
      <td>69.0</td>
      <td>22.0</td>
      <td>2.0</td>
      <td>0.233472</td>
      <td>0.094835</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3321136</th>
      <td>398372.0</td>
      <td>81.0</td>
      <td>104.0</td>
      <td>104.0</td>
      <td>153.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>48.5</td>
      <td>55.0</td>
      <td>55.0</td>
      <td>...</td>
      <td>70.0</td>
      <td>28.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>8.0</td>
      <td>3.0</td>
      <td>0.000000</td>
      <td>0.180938</td>
    </tr>
    <tr>
      <th>3321137</th>
      <td>398372.0</td>
      <td>81.0</td>
      <td>104.0</td>
      <td>104.0</td>
      <td>153.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>48.5</td>
      <td>55.0</td>
      <td>55.0</td>
      <td>...</td>
      <td>98.0</td>
      <td>31.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>8.0</td>
      <td>3.0</td>
      <td>0.000000</td>
      <td>0.203977</td>
    </tr>
    <tr>
      <th>3321138</th>
      <td>398372.0</td>
      <td>81.0</td>
      <td>104.0</td>
      <td>104.0</td>
      <td>153.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>48.5</td>
      <td>55.0</td>
      <td>55.0</td>
      <td>...</td>
      <td>3.0</td>
      <td>48.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>8.0</td>
      <td>3.0</td>
      <td>0.000000</td>
      <td>0.169803</td>
    </tr>
    <tr>
      <th>3321139</th>
      <td>398372.0</td>
      <td>81.0</td>
      <td>104.0</td>
      <td>104.0</td>
      <td>153.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>48.5</td>
      <td>55.0</td>
      <td>55.0</td>
      <td>...</td>
      <td>94.0</td>
      <td>32.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>8.0</td>
      <td>3.0</td>
      <td>0.000000</td>
      <td>0.120565</td>
    </tr>
    <tr>
      <th>3321140</th>
      <td>398372.0</td>
      <td>81.0</td>
      <td>104.0</td>
      <td>104.0</td>
      <td>153.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>48.5</td>
      <td>55.0</td>
      <td>55.0</td>
      <td>...</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>343.0</td>
      <td>39.0</td>
      <td>22.0</td>
      <td>79.0</td>
      <td>8.0</td>
      <td>3.0</td>
      <td>0.000000</td>
      <td>0.169962</td>
    </tr>
  </tbody>
</table>
<p>3321141 rows × 81 columns</p>
</div>
</div>
</div>
<p>This is the data we will use for training. Note that each spectrum comprises of multiple lines: One line per b/y-ion couple.</p>
</section>
</section>
<section id="training-the-model" class="level2">
<h2 class="anchored" data-anchor-id="training-the-model">3 Training the model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.ensemble <span class="im">import</span> GradientBoostingRegressor</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s first try to train a simple model and evaluate its performance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>reg <span class="op">=</span>  GradientBoostingRegressor()</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>X_train <span class="op">=</span> train_val_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>y_train <span class="op">=</span> train_val_encoded[<span class="st">"b_target"</span>]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>X_test <span class="op">=</span> test_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>y_test <span class="op">=</span> test_encoded[<span class="st">"b_target"</span>]</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>reg.fit(X_test, y_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>GradientBoostingRegressor()</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>y_test_pred <span class="op">=</span> reg.predict(X_test)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>np.corrcoef(y_test, y_test_pred)[<span class="dv">0</span>][<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0.696138939634552</code></pre>
</div>
</div>
<p>Not terrible. Let’s see if we can do better after hyperparameters optimization. For this, we can use the hyperopt package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> hyperopt <span class="im">import</span> fmin, hp, tpe, STATUS_OK</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> objective(n_estimators):</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Define algorithm</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    reg <span class="op">=</span>  GradientBoostingRegressor(n_estimators<span class="op">=</span>n_estimators)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Fit model</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    reg.fit(X_test, y_test)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Test model</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    y_test_pred <span class="op">=</span> reg.predict(X_test)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>    correlation <span class="op">=</span> np.corrcoef(y_test, y_test_pred)[<span class="dv">0</span>][<span class="dv">1</span>]</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {<span class="st">'loss'</span>: <span class="op">-</span>correlation, <span class="st">'status'</span>: STATUS_OK}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>best_params <span class="op">=</span> fmin(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  fn<span class="op">=</span>objective,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  space<span class="op">=</span><span class="dv">10</span> <span class="op">+</span> hp.randint(<span class="st">'n_estimators'</span>, <span class="dv">980</span>),</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  algo<span class="op">=</span>tpe.suggest,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  max_evals<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>best_params</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'n_estimators': 946}</code></pre>
</div>
</div>
<p>Initially, the default value of 100 estimators was used. According to this hyperopt run, using 946 estimators results in a more performant model.</p>
<p>Now we can train the model again with this new hyperparameter value:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>reg <span class="op">=</span>  GradientBoostingRegressor(n_estimators<span class="op">=</span><span class="dv">946</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>X_train <span class="op">=</span> train_val_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>y_train <span class="op">=</span> train_val_encoded[<span class="st">"b_target"</span>]</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>X_test <span class="op">=</span> test_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>y_test <span class="op">=</span> test_encoded[<span class="st">"b_target"</span>]</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>reg.fit(X_test, y_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>GradientBoostingRegressor(n_estimators=946)</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>y_test_pred <span class="op">=</span> reg.predict(X_test)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>np.corrcoef(y_test, y_test_pred)[<span class="dv">0</span>][<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0.7951796228632316</code></pre>
</div>
</div>
<p>Much better already. To get a more accurate view of the model performance, we should calculate the correlation per spectrum, instead of across the full dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>prediction_df_y <span class="op">=</span> pd.DataFrame({</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"spectrum_id"</span>: test_encoded[<span class="st">"spectrum_id"</span>], </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"target_y"</span>: y_test,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"prediction_y"</span>: y_test_pred,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>prediction_df_y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>spectrum_id</th>
      <th>target_y</th>
      <th>prediction_y</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>9.0</td>
      <td>0.000000</td>
      <td>0.036654</td>
    </tr>
    <tr>
      <th>1</th>
      <td>9.0</td>
      <td>0.104444</td>
      <td>0.036654</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9.0</td>
      <td>0.153251</td>
      <td>0.036654</td>
    </tr>
    <tr>
      <th>3</th>
      <td>9.0</td>
      <td>0.155234</td>
      <td>0.036654</td>
    </tr>
    <tr>
      <th>4</th>
      <td>9.0</td>
      <td>0.158532</td>
      <td>0.036910</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>367683</th>
      <td>398369.0</td>
      <td>0.018268</td>
      <td>-0.024844</td>
    </tr>
    <tr>
      <th>367684</th>
      <td>398369.0</td>
      <td>0.076188</td>
      <td>-0.036396</td>
    </tr>
    <tr>
      <th>367685</th>
      <td>398369.0</td>
      <td>0.000000</td>
      <td>-0.044847</td>
    </tr>
    <tr>
      <th>367686</th>
      <td>398369.0</td>
      <td>0.000000</td>
      <td>-0.061128</td>
    </tr>
    <tr>
      <th>367687</th>
      <td>398369.0</td>
      <td>0.000000</td>
      <td>-0.052229</td>
    </tr>
  </tbody>
</table>
<p>367688 rows × 3 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>corr_y <span class="op">=</span> prediction_df_y.groupby(<span class="st">"spectrum_id"</span>).corr().iloc[::<span class="dv">2</span>][<span class="st">'prediction_y'</span>]</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>corr_y.index <span class="op">=</span> corr_y.index.droplevel(<span class="dv">1</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>corr_y <span class="op">=</span> corr_y.reset_index().rename(columns<span class="op">=</span>{<span class="st">"prediction_y"</span>: <span class="st">"correlation"</span>})</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>corr_y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>spectrum_id</th>
      <th>correlation</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>9.0</td>
      <td>0.688671</td>
    </tr>
    <tr>
      <th>1</th>
      <td>16.0</td>
      <td>0.516480</td>
    </tr>
    <tr>
      <th>2</th>
      <td>39.0</td>
      <td>0.598204</td>
    </tr>
    <tr>
      <th>3</th>
      <td>95.0</td>
      <td>0.427736</td>
    </tr>
    <tr>
      <th>4</th>
      <td>140.0</td>
      <td>0.662748</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>27031</th>
      <td>398328.0</td>
      <td>0.545729</td>
    </tr>
    <tr>
      <th>27032</th>
      <td>398341.0</td>
      <td>0.327869</td>
    </tr>
    <tr>
      <th>27033</th>
      <td>398342.0</td>
      <td>0.400296</td>
    </tr>
    <tr>
      <th>27034</th>
      <td>398368.0</td>
      <td>0.329486</td>
    </tr>
    <tr>
      <th>27035</th>
      <td>398369.0</td>
      <td>0.496507</td>
    </tr>
  </tbody>
</table>
<p>27036 rows × 2 columns</p>
</div>
</div>
</div>
<p>Median correlation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>corr_y[<span class="st">"correlation"</span>].median()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0.8528908977025409</code></pre>
</div>
</div>
<p>Correlation distribution:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>sns.set_style(<span class="st">"whitegrid"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>sns.catplot(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>corr_y, x<span class="op">=</span><span class="st">"correlation"</span>,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    fliersize<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    kind<span class="op">=</span><span class="st">"box"</span>, aspect<span class="op">=</span><span class="dv">4</span>, height<span class="op">=</span><span class="dv">2</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="nist-2-traditional-ml-gradient-boosting_files/figure-html/cell-29-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Not bad! With some more hyperparameter optimization (optimizing only the number of trees is a bit crude) a lot more performance gains could be made. Take a look at the <a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html">Scikit Learn documentation</a> to learn more about the various hyperparameters for the <code>GradientBoostingRegressor</code>. Alternatively, you could switch to the <a href="https://xgboost.readthedocs.io/en/stable/">XGBoost algorithm</a>, which is currently used by MS²PIP.</p>
<p>And of course, this model can only predict y-ion intensities. You can repeat the training and optimization steps to train a model for b-ion intensities.</p>
<p>Good luck!</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Friedman2002" class="csl-entry" role="doc-biblioentry">
Friedman, Jerome H. 2002. <span>“Stochastic Gradient Boosting.”</span> <em>Computational Statistics and Data Analysis</em> 38 (4): 367–78. <a href="https://doi.org/10.1016/s0167-9473(01)00065-2">https://doi.org/10.1016/s0167-9473(01)00065-2</a>.
</div>
<div id="ref-Gabriels2019" class="csl-entry" role="doc-biblioentry">
Gabriels, Ralf, Lennart Martens, and Sven Degroeve. 2019. <span>“Updated MS²PIP Web Server Delivers Fast and Accurate MS² Peak Intensity Prediction for Multiple Fragmentation Methods, Instruments and Labeling Techniques.”</span> <em>NUCLEIC ACIDS RESEARCH</em> 47 (W1): W295–99. <a href="https://doi.org/10.1093/nar/gkz299">https://doi.org/10.1093/nar/gkz299</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp(/https:\/\/www\.proteomicsml\.org/);
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
    var links = window.document.querySelectorAll('a:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center"><small>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons CC BY 4.0</a> license.</small></div>
  </div>
</footer>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>