<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ralf Gabriels">
<meta name="dcterms.date" content="2022-10-05">

<title>ProteomicsML - NIST (part 2): Traditional ML: Gradient boosting</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../static/img/proteomicsml-icon.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-VVRMZHT2W2"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-VVRMZHT2W2', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"interstitial",
  "consent_type":"express",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  });
});
</script> 
  
<style>html{ scroll-behavior: smooth; }</style>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "LearningResource",
    "@id": "https://proteomicsml.org/",
    "dct:conformsTo": {
      "http://purl.org/dc/terms/conformsTo": {
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
        "@type": "CreativeWork"
      }
    },
    "name": "ProteomicsML",
    "description": "ProteomicsML provides ready-made datasets for machine learning models accompanied by tutorials on how to work with even the most complex data types in the field of proteomics. The resource is set up to evolve together with the field, and we welcome everyone to contribute to the project by adding new datasets and accompanying notebooks.",
    "keywords": [
      "machine learning",
      "deep learning",
      "proteomics",
      "educational platform",
      "community platform",
      "bioinformatics",
      "detectability",
      "fragmentation",
      "retention time",
      "ion mobility"
    ],
    "about": [
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/topic_0091",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_0091",
        "name": "bioinformatics",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Ftopic_0091"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/3474",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_3474",
        "name": "machine learning",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Ftopic_3474"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/0121",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_0121",
        "name": "proteomics",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Ftopic_0121"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/data_3670",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "data_3670",
        "name": "Online course",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fdata_3670"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/data_2536",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "data_2536",
        "name": "Mass spectrometry data",
        "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fdata_2536"
      }
    ],
    "abstract": "Data set acquisition and curation are often the most difficult and time-consuming parts of a machine learning endeavor. This is especially true for proteomics-based liquid chromatography (LC) coupled to mass spectrometry (MS) data sets, due to the high levels of data reduction that occur between raw data and machine learning-ready data. Since predictive proteomics is an emerging field, when predicting peptide behavior in LC-MS setups, each lab often uses unique and complex data processing pipelines in order to maximize performance, at the cost of accessibility and reproducibility. For this reason we introduce ProteomicsML, an online resource for proteomics-based data sets and tutorials across most of the currently explored physicochemical peptide properties. This community-driven resource makes it simple to access data in easy-to-process formats, and contains easy-to-follow tutorials that allow new users to interact with even the most advanced algorithms in the field. ProteomicsML provides data sets that are useful for comparing state-of-the-art machine learning algorithms, as well as providing introductory material for teachers and newcomers to the field alike. The platform is freely available at https://www.proteomicsml.org/, and we welcome the entire proteomics community to contribute to the project at https://github.com/ProteomicsML/ProteomicsML.",
    "audience": [
      {
        "@type": "Audience",
        "audienceType": "Students"
      },
      {
        "@type": "Audience",
        "audienceType": "PhD students"
      },
      {
        "@type": "Audience",
        "audienceType": "Postdoctoral researchers"
      }
    ],
    "competencyRequired": [
      "Basic python programming",
      "Machine learning basics",
      "Basic proteomics and mass spectrometry knowledge"
    ],
    "inLanguage": ["en-US"],
    "learningResourceType": ["tutorials"],
    "license": "https://spdx.org/licenses/CC-BY-4.0.html",
    "url": "https://proteomicsml.org/"
  }
</script>


<link rel="stylesheet" href="../../static/css/custom-styles.css">
<meta property="og:title" content="ProteomicsML - NIST (part 2): Traditional ML: Gradient boosting">
<meta property="og:description" content="">
<meta property="og:image" content="https://www.proteomicsml.org/tutorials/fragmentation/data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==">
<meta property="og:site-name" content="ProteomicsML">
<meta name="twitter:title" content="ProteomicsML - NIST (part 2): Traditional ML: Gradient boosting">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://www.proteomicsml.org/tutorials/fragmentation/data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../static/img/proteomicsml-logo-inverse.svg" alt="ProteomicsML logo" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../publication.html" rel="" target="">
 <span class="menu-text">Publication</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../tutorials/index.html" rel="" target="">
 <span class="menu-text">Tutorials</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../datasets/index.html" rel="" target="">
 <span class="menu-text">Datasets</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../contributing.html" rel="" target="">
 <span class="menu-text">Contributing</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../code-of-conduct.html" rel="" target="">
 <span class="menu-text">Code of Conduct</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ProteomicsML" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/ProteomicsML" rel="" target=""><i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../tutorials/index.html">Tutorials</a></li><li class="breadcrumb-item"><a href="../../tutorials/fragmentation/index.html">Fragmentation</a></li><li class="breadcrumb-item"><a href="../../tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.html">NIST (part 2): Traditional ML: Gradient boosting</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../publication.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Publication</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../tutorials/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tutorials</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../tutorials/detectability/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Detectability</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/detectability/modeling-protein-detectability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Modelling protein detectability with an MLP</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../tutorials/retentiontime/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Retention time</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/deeplc-transfer-learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transfer learning with DeepLC</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/dlomix-prosit-rt.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DLOmix embedding of Prosit model on ProteomeTools data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/manual-prosit-rt.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Manual embedding of Bi-LSTM model on ProteomeTools data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/retentiontime/mq-evidence-to-ml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preparing a retention time data set for machine learning</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../tutorials/fragmentation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Fragmentation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/nist-1-parsing-spectral-library.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">NIST (part 1): Preparing a spectral library for ML</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">NIST (part 2): Traditional ML: Gradient boosting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/preannotated-prosit.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prosit-style GRU with pre-annotated ProteomeTools data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/fragmentation/raw-to-prosit.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Raw file processing with PROSIT style annotation</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../tutorials/ionmobility/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ion mobility</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../tutorials/ionmobility/meier-tims-ccs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Predicting CCS values for TIMS data</span></a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../datasets/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false">
 <span class="menu-text">Detectability</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/detectability/ArabidopsisLightDarkProteome.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Arabidopsis PeptideAtlas Light and Dark Proteome</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="false">
 <span class="menu-text">Retention time</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/retentiontime/DLOmix_RT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DLOmix</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/retentiontime/ProteomeTools_RT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ProteomeTools</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/retentiontime/PXD028248_RT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">PXD028248</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="false">
 <span class="menu-text">Fragmentation</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/fragmentation/nist.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">NIST Peptide libraries</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/fragmentation/ProteomeTools_FI.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ProteomeTools</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" aria-expanded="false">
 <span class="menu-text">Ion mobility</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-10" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/ionmobility/Meier_TIMS.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Meier et al.&nbsp;TIMS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../datasets/ionmobility/VanPuyvelde_TWIMS.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Van Puyvelde et al.&nbsp;TWIMS</span></a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../contributing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Contributing</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../code-of-conduct.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Code of Conduct</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">1. Introduction</a></li>
  <li><a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation">2 Data preparation</a>
  <ul>
  <li><a href="#feature-engineering" id="toc-feature-engineering" class="nav-link" data-scroll-target="#feature-engineering">2.1 Feature engineering</a></li>
  <li><a href="#getting-the-target-intensities" id="toc-getting-the-target-intensities" class="nav-link" data-scroll-target="#getting-the-target-intensities">2.2 Getting the target intensities</a></li>
  <li><a href="#bringing-it-all-together" id="toc-bringing-it-all-together" class="nav-link" data-scroll-target="#bringing-it-all-together">2.3 Bringing it all together</a></li>
  </ul></li>
  <li><a href="#training-the-model" id="toc-training-the-model" class="nav-link" data-scroll-target="#training-the-model">3 Training the model</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ProteomicsML/ProteomicsML/blob/main/tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.ipynb" class="toc-action">View source</a></p><p><a href="https://github.dev/ProteomicsML/ProteomicsML/blob/main/tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.ipynb" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">NIST (part 2): Traditional ML: Gradient boosting</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Ralf Gabriels <a href="https://orcid.org/0000-0002-1679-1711" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            VIB-UGent Center for Medical Biotechnology, VIB, Belgium
          </p>
        <p class="affiliation">
            Department of Biomolecular Medicine, Ghent University, Belgium
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 5, 2022</p>
    </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Modified</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">August 28, 2025</p>
    </div>
  </div>
    
  </div>
  

</header>

<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://colab.research.google.com/github/ProteomicsML/ProteomicsML/blob/main/tutorials/fragmentation/_nist-2-traditional-ml-gradient-boosting.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" class="img-fluid figure-img"></a></p>
</figure>
</div>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">1. Introduction</h2>
<p>This is the second part in a three-part tutorial. We recommend you to to start with the first section, where the NIST spectral library is parsed and prepared for use in the second and third parts.</p>
<ol type="1">
<li><a href="https://www.proteomicsml.org/tutorials/fragmentation/nist-1-parsing-spectral-library.html">Preparing a spectral library for ML</a></li>
<li><a href="https://www.proteomicsml.org/tutorials/fragmentation/nist-2-traditional-ml-gradient-boosting.html">Traditional ML: Gradient boosting</a></li>
<li><a href="https://www.proteomicsml.org/tutorials/fragmentation/nist-3-deep-learning-lstm.html">Deep learning: BiLSTM</a></li>
</ol>
<p>In this tutorial, you will learn how to build a fragmentation intensity predictor similar to MS²PIP v3 <span class="citation" data-cites="Gabriels2019">(<a href="#ref-Gabriels2019" role="doc-biblioref">Gabriels, Martens, and Degroeve 2019</a>)</span> with traditional machine learning (ML) feature engineering and Gradient Boosting <span class="citation" data-cites="Friedman2002">(<a href="#ref-Friedman2002" role="doc-biblioref">Friedman 2002</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Installing required python packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> pip install rich numpy pandas pyarrow matplotlib seaborn scikit<span class="op">-</span>learn hyperopt <span class="op">--</span>quiet</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-preparation" class="level2">
<h2 class="anchored" data-anchor-id="data-preparation">2 Data preparation</h2>
<p>We will use the spectral library that was already parsed in part 1 of this tutorial series.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>train_val_spectra <span class="op">=</span> pd.read_feather(<span class="st">"http://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-trainval.feather"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>test_spectra <span class="op">=</span> pd.read_feather(<span class="st">"http://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-test.feather"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="feature-engineering" class="level3">
<h3 class="anchored" data-anchor-id="feature-engineering">2.1 Feature engineering</h3>
<p>In traditional ML, the input features for the algorithm usually require some engineering. For fragmentation intensity prediction this is not different. Following the MS²PIP methods, we will calculate the distributions of several amino acid properties across the peptide and fragment ion sequences.</p>
<p>Using the distribution of these properties instead of the actual properties per amino acid allows MS²PIP to get a fixed length feature matrix for input peptides with varying lengths.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> rich <span class="im">import</span> progress</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>amino_acids <span class="op">=</span> <span class="bu">list</span>(<span class="st">"ACDEFGHIKLMNPQRSTVWY"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>properties <span class="op">=</span> np.array([</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">37</span>,<span class="dv">35</span>,<span class="dv">59</span>,<span class="dv">129</span>,<span class="dv">94</span>,<span class="dv">0</span>,<span class="dv">210</span>,<span class="dv">81</span>,<span class="dv">191</span>,<span class="dv">81</span>,<span class="dv">106</span>,<span class="dv">101</span>,<span class="dv">117</span>,<span class="dv">115</span>,<span class="dv">343</span>,<span class="dv">49</span>,<span class="dv">90</span>,<span class="dv">60</span>,<span class="dv">134</span>,<span class="dv">104</span>],  <span class="co"># basicity</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">68</span>,<span class="dv">23</span>,<span class="dv">33</span>,<span class="dv">29</span>,<span class="dv">70</span>,<span class="dv">58</span>,<span class="dv">41</span>,<span class="dv">73</span>,<span class="dv">32</span>,<span class="dv">73</span>,<span class="dv">66</span>,<span class="dv">38</span>,<span class="dv">0</span>,<span class="dv">40</span>,<span class="dv">39</span>,<span class="dv">44</span>,<span class="dv">53</span>,<span class="dv">71</span>,<span class="dv">51</span>,<span class="dv">55</span>],  <span class="co"># helicity</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">51</span>,<span class="dv">75</span>,<span class="dv">25</span>,<span class="dv">35</span>,<span class="dv">100</span>,<span class="dv">16</span>,<span class="dv">3</span>,<span class="dv">94</span>,<span class="dv">0</span>,<span class="dv">94</span>,<span class="dv">82</span>,<span class="dv">12</span>,<span class="dv">0</span>,<span class="dv">22</span>,<span class="dv">22</span>,<span class="dv">21</span>,<span class="dv">39</span>,<span class="dv">80</span>,<span class="dv">98</span>,<span class="dv">70</span>],  <span class="co"># hydrophobicity</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">32</span>,<span class="dv">23</span>,<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">27</span>,<span class="dv">32</span>,<span class="dv">48</span>,<span class="dv">32</span>,<span class="dv">69</span>,<span class="dv">32</span>,<span class="dv">29</span>,<span class="dv">26</span>,<span class="dv">35</span>,<span class="dv">28</span>,<span class="dv">79</span>,<span class="dv">29</span>,<span class="dv">28</span>,<span class="dv">31</span>,<span class="dv">31</span>,<span class="dv">28</span>],  <span class="co"># pI</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>])</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(properties, columns<span class="op">=</span>amino_acids, index<span class="op">=</span>[<span class="st">"basicity"</span>, <span class="st">"helicity"</span>, <span class="st">"hydrophobicity"</span>, <span class="st">"pI"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">D</th>
<th data-quarto-table-cell-role="th">E</th>
<th data-quarto-table-cell-role="th">F</th>
<th data-quarto-table-cell-role="th">G</th>
<th data-quarto-table-cell-role="th">H</th>
<th data-quarto-table-cell-role="th">I</th>
<th data-quarto-table-cell-role="th">K</th>
<th data-quarto-table-cell-role="th">L</th>
<th data-quarto-table-cell-role="th">M</th>
<th data-quarto-table-cell-role="th">N</th>
<th data-quarto-table-cell-role="th">P</th>
<th data-quarto-table-cell-role="th">Q</th>
<th data-quarto-table-cell-role="th">R</th>
<th data-quarto-table-cell-role="th">S</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">V</th>
<th data-quarto-table-cell-role="th">W</th>
<th data-quarto-table-cell-role="th">Y</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">basicity</td>
<td>37</td>
<td>35</td>
<td>59</td>
<td>129</td>
<td>94</td>
<td>0</td>
<td>210</td>
<td>81</td>
<td>191</td>
<td>81</td>
<td>106</td>
<td>101</td>
<td>117</td>
<td>115</td>
<td>343</td>
<td>49</td>
<td>90</td>
<td>60</td>
<td>134</td>
<td>104</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">helicity</td>
<td>68</td>
<td>23</td>
<td>33</td>
<td>29</td>
<td>70</td>
<td>58</td>
<td>41</td>
<td>73</td>
<td>32</td>
<td>73</td>
<td>66</td>
<td>38</td>
<td>0</td>
<td>40</td>
<td>39</td>
<td>44</td>
<td>53</td>
<td>71</td>
<td>51</td>
<td>55</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">hydrophobicity</td>
<td>51</td>
<td>75</td>
<td>25</td>
<td>35</td>
<td>100</td>
<td>16</td>
<td>3</td>
<td>94</td>
<td>0</td>
<td>94</td>
<td>82</td>
<td>12</td>
<td>0</td>
<td>22</td>
<td>22</td>
<td>21</td>
<td>39</td>
<td>80</td>
<td>98</td>
<td>70</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">pI</td>
<td>32</td>
<td>23</td>
<td>0</td>
<td>4</td>
<td>27</td>
<td>32</td>
<td>48</td>
<td>32</td>
<td>69</td>
<td>32</td>
<td>29</td>
<td>26</td>
<td>35</td>
<td>28</td>
<td>79</td>
<td>29</td>
<td>28</td>
<td>31</td>
<td>31</td>
<td>28</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> encode_peptide(sequence, charge):</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 4 properties * 5 quantiles * 3 ion types + 4 properties * 4 site + 2 global</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    n_features <span class="op">=</span> <span class="dv">78</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    quantiles <span class="op">=</span> [<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="dv">1</span>]</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    n_ions <span class="op">=</span> <span class="bu">len</span>(sequence) <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Encode amino acids as integers to index amino acid properties for peptide sequence</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    aa_indices <span class="op">=</span> {aa: i <span class="cf">for</span> i, aa <span class="kw">in</span>  <span class="bu">enumerate</span>(<span class="st">"ACDEFGHIKLMNPQRSTVWY"</span>)}</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    aa_to_index <span class="op">=</span> np.vectorize(<span class="kw">lambda</span> aa: aa_indices[aa])</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    peptide_indexed <span class="op">=</span> aa_to_index(np.array(<span class="bu">list</span>(sequence)))</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    peptide_properties <span class="op">=</span> properties[:, peptide_indexed]</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Empty peptide_features array</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    peptide_features <span class="op">=</span> np.full((n_ions, n_features), np.nan)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> b_ion_number <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>, n_ions <span class="op">+</span> <span class="dv">1</span>):</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate quantiles of features across peptide, b-ion, and y-ion</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>        peptide_quantiles <span class="op">=</span> np.hstack(</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>            np.quantile(peptide_properties, quantiles, axis<span class="op">=</span><span class="dv">1</span>).transpose()</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>        b_ion_quantiles <span class="op">=</span> np.hstack(</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>            np.quantile(peptide_properties[:,:b_ion_number], quantiles, axis<span class="op">=</span><span class="dv">1</span>).transpose()</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>        y_ion_quantiles <span class="op">=</span> np.hstack(</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>            np.quantile(peptide_properties[:,b_ion_number:], quantiles, axis<span class="op">=</span><span class="dv">1</span>).transpose()</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Properties on specific sites: nterm, frag-1, frag+1, cterm</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>        specific_site_indexes <span class="op">=</span> np.array([<span class="dv">0</span>, b_ion_number <span class="op">-</span> <span class="dv">1</span>, b_ion_number, <span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>        specific_site_properties <span class="op">=</span> np.hstack(peptide_properties[:, specific_site_indexes].transpose())</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Global features: Length and charge</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>        global_features <span class="op">=</span> np.array([<span class="bu">len</span>(sequence), <span class="bu">int</span>(charge)])</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Assign to peptide_features array</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">0</span>:<span class="dv">20</span>] <span class="op">=</span> peptide_quantiles</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">20</span>:<span class="dv">40</span>] <span class="op">=</span> b_ion_quantiles</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">40</span>:<span class="dv">60</span>] <span class="op">=</span> y_ion_quantiles</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">60</span>:<span class="dv">76</span>] <span class="op">=</span> specific_site_properties</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>        peptide_features[b_ion_number <span class="op">-</span> <span class="dv">1</span>, <span class="dv">76</span>:<span class="dv">78</span>] <span class="op">=</span> global_features</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> peptide_features</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_feature_names():</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>    feature_names <span class="op">=</span> []</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> level <span class="kw">in</span> [<span class="st">"peptide"</span>, <span class="st">"b"</span>, <span class="st">"y"</span>]:</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> aa_property <span class="kw">in</span> [<span class="st">"basicity"</span>, <span class="st">"helicity"</span>, <span class="st">"hydrophobicity"</span>, <span class="st">"pi"</span>]:</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> quantile <span class="kw">in</span> [<span class="st">"min"</span>, <span class="st">"q1"</span>, <span class="st">"q2"</span>, <span class="st">"q3"</span>, <span class="st">"max"</span>]:</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>                feature_names.append(<span class="st">"_"</span>.join([level, aa_property, quantile]))</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> site <span class="kw">in</span> [<span class="st">"nterm"</span>, <span class="st">"fragmin1"</span>, <span class="st">"fragplus1"</span>, <span class="st">"cterm"</span>]:</span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> aa_property <span class="kw">in</span> [<span class="st">"basicity"</span>, <span class="st">"helicity"</span>, <span class="st">"hydrophobicity"</span>, <span class="st">"pi"</span>]:</span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>            feature_names.append(<span class="st">"_"</span>.join([site, aa_property]))</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>    feature_names.extend([<span class="st">"length"</span>, <span class="st">"charge"</span>])</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> feature_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s test it with a single peptide. Feel free to use your own name as a “peptide”; as long as it does not contain any non-amino acid characters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>peptide_features <span class="op">=</span> pd.DataFrame(encode_peptide(<span class="st">"RALFGARIELS"</span>, <span class="dv">2</span>), columns<span class="op">=</span>generate_feature_names())</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>peptide_features</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">peptide_basicity_min</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q1</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q2</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q3</th>
<th data-quarto-table-cell-role="th">peptide_basicity_max</th>
<th data-quarto-table-cell-role="th">peptide_helicity_min</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q1</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q2</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q3</th>
<th data-quarto-table-cell-role="th">peptide_helicity_max</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">fragplus1_basicity</th>
<th data-quarto-table-cell-role="th">fragplus1_helicity</th>
<th data-quarto-table-cell-role="th">fragplus1_hydrophobicity</th>
<th data-quarto-table-cell-role="th">fragplus1_pi</th>
<th data-quarto-table-cell-role="th">cterm_basicity</th>
<th data-quarto-table-cell-role="th">cterm_helicity</th>
<th data-quarto-table-cell-role="th">cterm_hydrophobicity</th>
<th data-quarto-table-cell-role="th">cterm_pi</th>
<th data-quarto-table-cell-role="th">length</th>
<th data-quarto-table-cell-role="th">charge</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>37.0</td>
<td>68.0</td>
<td>51.0</td>
<td>32.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>81.0</td>
<td>73.0</td>
<td>94.0</td>
<td>32.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>94.0</td>
<td>70.0</td>
<td>100.0</td>
<td>27.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>0.0</td>
<td>58.0</td>
<td>16.0</td>
<td>32.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>37.0</td>
<td>68.0</td>
<td>51.0</td>
<td>32.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>81.0</td>
<td>73.0</td>
<td>94.0</td>
<td>32.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>129.0</td>
<td>29.0</td>
<td>35.0</td>
<td>4.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>81.0</td>
<td>73.0</td>
<td>94.0</td>
<td>32.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>0.0</td>
<td>43.0</td>
<td>81.0</td>
<td>111.5</td>
<td>343.0</td>
<td>29.0</td>
<td>41.5</td>
<td>68.0</td>
<td>71.5</td>
<td>73.0</td>
<td>...</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>49.0</td>
<td>44.0</td>
<td>21.0</td>
<td>29.0</td>
<td>11.0</td>
<td>2.0</td>
</tr>
</tbody>
</table>

<p>10 rows × 78 columns</p>
</div>
</div>
</div>
</section>
<section id="getting-the-target-intensities" class="level3">
<h3 class="anchored" data-anchor-id="getting-the-target-intensities">2.2 Getting the target intensities</h3>
<p>The target intensities are the observed intensities which the model will learn to predict. Let’s first try with a single spectrum.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>test_spectrum <span class="op">=</span> train_val_spectra.iloc[<span class="dv">4</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>peptide_targets <span class="op">=</span>  pd.DataFrame({</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"b_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>],</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"y_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>],</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>peptide_targets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">b_target</th>
<th data-quarto-table-cell-role="th">y_target</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.000000</td>
<td>0.118507</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.229717</td>
<td>0.079770</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.294631</td>
<td>0.088712</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.234662</td>
<td>0.145900</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>0.185732</td>
<td>0.205005</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>0.134395</td>
<td>0.261630</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>0.081856</td>
<td>0.305119</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>0.043793</td>
<td>0.296351</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>0.000000</td>
<td>0.205703</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>0.000000</td>
<td>0.155991</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>These are the intensities for the b- and y-ions, each ordered from 1 to 9. In MS²PIP, however, a clever trick is applied to reuse the computed features for each <em>fragment ion pair</em>. Doing so makes perfect sense, as both ions in such a fragment ion pair originated from the same fragmentation event. For this peptide, the fragment ion pairs are b1-y9, b2-y8, b3-y7, etc. To match all of the pairs, we simply have to reverse the y-ion series intensities:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>peptide_targets <span class="op">=</span>  pd.DataFrame({</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"b_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>],</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"y_target"</span>: test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>][::<span class="op">-</span><span class="dv">1</span>],</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>peptide_targets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">b_target</th>
<th data-quarto-table-cell-role="th">y_target</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.229717</td>
<td>0.155991</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.294631</td>
<td>0.205703</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.234662</td>
<td>0.296351</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>0.185732</td>
<td>0.305119</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>0.134395</td>
<td>0.261630</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>0.081856</td>
<td>0.205005</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>0.043793</td>
<td>0.145900</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>0.000000</td>
<td>0.088712</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>0.000000</td>
<td>0.079770</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>0.000000</td>
<td>0.118507</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="bringing-it-all-together" class="level3">
<h3 class="anchored" data-anchor-id="bringing-it-all-together">2.3 Bringing it all together</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>features <span class="op">=</span> encode_peptide(test_spectrum[<span class="st">"sequence"</span>], test_spectrum[<span class="st">"charge"</span>])</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>targets <span class="op">=</span> np.stack([test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>], test_spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>][::<span class="op">-</span><span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>spectrum_id <span class="op">=</span> np.full(shape<span class="op">=</span>(targets.shape[<span class="dv">0</span>], <span class="dv">1</span>), fill_value<span class="op">=</span>test_spectrum[<span class="st">"index"</span>])  <span class="co"># Repeat id for all ions</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(np.hstack([spectrum_id, features, targets]), columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>] <span class="op">+</span> generate_feature_names() <span class="op">+</span> [<span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">spectrum_id</th>
<th data-quarto-table-cell-role="th">peptide_basicity_min</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q1</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q2</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q3</th>
<th data-quarto-table-cell-role="th">peptide_basicity_max</th>
<th data-quarto-table-cell-role="th">peptide_helicity_min</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q1</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q2</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q3</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">fragplus1_hydrophobicity</th>
<th data-quarto-table-cell-role="th">fragplus1_pi</th>
<th data-quarto-table-cell-role="th">cterm_basicity</th>
<th data-quarto-table-cell-role="th">cterm_helicity</th>
<th data-quarto-table-cell-role="th">cterm_hydrophobicity</th>
<th data-quarto-table-cell-role="th">cterm_pi</th>
<th data-quarto-table-cell-role="th">length</th>
<th data-quarto-table-cell-role="th">charge</th>
<th data-quarto-table-cell-role="th">b_target</th>
<th data-quarto-table-cell-role="th">y_target</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.229717</td>
<td>0.155991</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.294631</td>
<td>0.205703</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.234662</td>
<td>0.296351</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.185732</td>
<td>0.305119</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.134395</td>
<td>0.261630</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.081856</td>
<td>0.205005</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.043793</td>
<td>0.145900</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>80.0</td>
<td>31.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.000000</td>
<td>0.088712</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>21.0</td>
<td>29.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.000000</td>
<td>0.079770</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>5.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>40.0</td>
<td>343.0</td>
<td>39.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>22.0</td>
<td>79.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>12.0</td>
<td>2.0</td>
<td>0.000000</td>
<td>0.118507</td>
</tr>
</tbody>
</table>

<p>11 rows × 81 columns</p>
</div>
</div>
</div>
<p>The following function applies these steps over a collection of spectra and returns the full feature/target table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_ml_input(spectra):</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    tables <span class="op">=</span> []</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> spectrum <span class="kw">in</span> progress.track(spectra.to_dict(orient<span class="op">=</span><span class="st">"records"</span>)):</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>        features <span class="op">=</span> encode_peptide(spectrum[<span class="st">"sequence"</span>], spectrum[<span class="st">"charge"</span>])</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>        targets <span class="op">=</span> np.stack([spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"b"</span>], spectrum[<span class="st">"parsed_intensity"</span>][<span class="st">"y"</span>][::<span class="op">-</span><span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>        spectrum_id <span class="op">=</span> np.full(shape<span class="op">=</span>(targets.shape[<span class="dv">0</span>], <span class="dv">1</span>), fill_value<span class="op">=</span>spectrum[<span class="st">"index"</span>])  <span class="co"># Repeat id for all ions</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>        table <span class="op">=</span> np.hstack([spectrum_id, features, targets])</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>        tables.append(table)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    full_table <span class="op">=</span> np.vstack(tables)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    spectra_encoded <span class="op">=</span> pd.DataFrame(full_table, columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>] <span class="op">+</span> generate_feature_names() <span class="op">+</span> [<span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> spectra_encoded</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this might take some time, sometimes up to 30 minutes. To skip this step, simple download the file with pre-encoded features and targets, and load in two cells below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>train_val_encoded <span class="op">=</span> generate_ml_input(train_val_spectra)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>train_val_encoded.to_feather(<span class="st">"fragmentation-nist-humanhcd20160503-parsed-trainval-encoded.feather"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>test_encoded <span class="op">=</span> generate_ml_input(test_spectra)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>test_encoded.to_feather(<span class="st">"fragmentation-nist-humanhcd20160503-parsed-test-encoded.feather"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uncomment this step to load pre-encoded features from a file:</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># train_val_encoded = pd.read_feather("http://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-trainval-encoded.feather")</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># test_encoded = pd.read_feather("http://ftp.pride.ebi.ac.uk/pub/databases/pride/resources/proteomicsml/fragmentation/nist-humanhcd20160503-parsed-test-encoded.feather")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>train_val_encoded</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">spectrum_id</th>
<th data-quarto-table-cell-role="th">peptide_basicity_min</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q1</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q2</th>
<th data-quarto-table-cell-role="th">peptide_basicity_q3</th>
<th data-quarto-table-cell-role="th">peptide_basicity_max</th>
<th data-quarto-table-cell-role="th">peptide_helicity_min</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q1</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q2</th>
<th data-quarto-table-cell-role="th">peptide_helicity_q3</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">fragplus1_hydrophobicity</th>
<th data-quarto-table-cell-role="th">fragplus1_pi</th>
<th data-quarto-table-cell-role="th">cterm_basicity</th>
<th data-quarto-table-cell-role="th">cterm_helicity</th>
<th data-quarto-table-cell-role="th">cterm_hydrophobicity</th>
<th data-quarto-table-cell-role="th">cterm_pi</th>
<th data-quarto-table-cell-role="th">length</th>
<th data-quarto-table-cell-role="th">charge</th>
<th data-quarto-table-cell-role="th">b_target</th>
<th data-quarto-table-cell-role="th">y_target</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.0</td>
<td>0.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>191.0</td>
<td>32.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>191.0</td>
<td>32.0</td>
<td>0.0</td>
<td>69.0</td>
<td>22.0</td>
<td>2.0</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.0</td>
<td>0.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>191.0</td>
<td>32.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>191.0</td>
<td>32.0</td>
<td>0.0</td>
<td>69.0</td>
<td>22.0</td>
<td>2.0</td>
<td>0.094060</td>
<td>0.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.0</td>
<td>0.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>191.0</td>
<td>32.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>191.0</td>
<td>32.0</td>
<td>0.0</td>
<td>69.0</td>
<td>22.0</td>
<td>2.0</td>
<td>0.180642</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.0</td>
<td>0.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>191.0</td>
<td>32.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>191.0</td>
<td>32.0</td>
<td>0.0</td>
<td>69.0</td>
<td>22.0</td>
<td>2.0</td>
<td>0.204203</td>
<td>0.050476</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>0.0</td>
<td>0.0</td>
<td>37.0</td>
<td>37.0</td>
<td>37.0</td>
<td>191.0</td>
<td>32.0</td>
<td>68.0</td>
<td>68.0</td>
<td>68.0</td>
<td>...</td>
<td>51.0</td>
<td>32.0</td>
<td>191.0</td>
<td>32.0</td>
<td>0.0</td>
<td>69.0</td>
<td>22.0</td>
<td>2.0</td>
<td>0.233472</td>
<td>0.094835</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3321136</td>
<td>398372.0</td>
<td>81.0</td>
<td>104.0</td>
<td>104.0</td>
<td>153.0</td>
<td>343.0</td>
<td>39.0</td>
<td>48.5</td>
<td>55.0</td>
<td>55.0</td>
<td>...</td>
<td>70.0</td>
<td>28.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>8.0</td>
<td>3.0</td>
<td>0.000000</td>
<td>0.180938</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3321137</td>
<td>398372.0</td>
<td>81.0</td>
<td>104.0</td>
<td>104.0</td>
<td>153.0</td>
<td>343.0</td>
<td>39.0</td>
<td>48.5</td>
<td>55.0</td>
<td>55.0</td>
<td>...</td>
<td>98.0</td>
<td>31.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>8.0</td>
<td>3.0</td>
<td>0.000000</td>
<td>0.203977</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3321138</td>
<td>398372.0</td>
<td>81.0</td>
<td>104.0</td>
<td>104.0</td>
<td>153.0</td>
<td>343.0</td>
<td>39.0</td>
<td>48.5</td>
<td>55.0</td>
<td>55.0</td>
<td>...</td>
<td>3.0</td>
<td>48.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>8.0</td>
<td>3.0</td>
<td>0.000000</td>
<td>0.169803</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3321139</td>
<td>398372.0</td>
<td>81.0</td>
<td>104.0</td>
<td>104.0</td>
<td>153.0</td>
<td>343.0</td>
<td>39.0</td>
<td>48.5</td>
<td>55.0</td>
<td>55.0</td>
<td>...</td>
<td>94.0</td>
<td>32.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>8.0</td>
<td>3.0</td>
<td>0.000000</td>
<td>0.120565</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3321140</td>
<td>398372.0</td>
<td>81.0</td>
<td>104.0</td>
<td>104.0</td>
<td>153.0</td>
<td>343.0</td>
<td>39.0</td>
<td>48.5</td>
<td>55.0</td>
<td>55.0</td>
<td>...</td>
<td>22.0</td>
<td>79.0</td>
<td>343.0</td>
<td>39.0</td>
<td>22.0</td>
<td>79.0</td>
<td>8.0</td>
<td>3.0</td>
<td>0.000000</td>
<td>0.169962</td>
</tr>
</tbody>
</table>

<p>3321141 rows × 81 columns</p>
</div>
</div>
</div>
<p>This is the data we will use for training. Note that each spectrum comprises of multiple lines: One line per b/y-ion couple.</p>
</section>
</section>
<section id="training-the-model" class="level2">
<h2 class="anchored" data-anchor-id="training-the-model">3 Training the model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.ensemble <span class="im">import</span> GradientBoostingRegressor</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s first try to train a simple model on the train set and evaluate its performance on the test set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>reg <span class="op">=</span>  GradientBoostingRegressor()</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>X_train <span class="op">=</span> train_val_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>y_train <span class="op">=</span> train_val_encoded[<span class="st">"y_target"</span>]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>X_test <span class="op">=</span> test_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>y_test <span class="op">=</span> test_encoded[<span class="st">"y_target"</span>]</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co"># For demo purposes, we only use the first 100k samples</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>reg.fit(X_train.head(<span class="dv">100000</span>), y_train.head(<span class="dv">100000</span>))</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Uncomment this step to train the model on the full dataset:</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="co"># reg.fit(X_train, y_train)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>#sk-container-id-2 {
  /* Definition of color scheme common for light and dark mode */
  --sklearn-color-text: black;
  --sklearn-color-line: gray;
  /* Definition of color scheme for unfitted estimators */
  --sklearn-color-unfitted-level-0: #fff5e6;
  --sklearn-color-unfitted-level-1: #f6e4d2;
  --sklearn-color-unfitted-level-2: #ffe0b3;
  --sklearn-color-unfitted-level-3: chocolate;
  /* Definition of color scheme for fitted estimators */
  --sklearn-color-fitted-level-0: #f0f8ff;
  --sklearn-color-fitted-level-1: #d4ebff;
  --sklearn-color-fitted-level-2: #b3dbfd;
  --sklearn-color-fitted-level-3: cornflowerblue;

  /* Specific color for light theme */
  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));
  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));
  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));
  --sklearn-color-icon: #696969;

  @media (prefers-color-scheme: dark) {
    /* Redefinition of color scheme for dark theme */
    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));
    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));
    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));
    --sklearn-color-icon: #878787;
  }
}

#sk-container-id-2 {
  color: var(--sklearn-color-text);
}

#sk-container-id-2 pre {
  padding: 0;
}

#sk-container-id-2 input.sk-hidden--visually {
  border: 0;
  clip: rect(1px 1px 1px 1px);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

#sk-container-id-2 div.sk-dashed-wrapped {
  border: 1px dashed var(--sklearn-color-line);
  margin: 0 0.4em 0.5em 0.4em;
  box-sizing: border-box;
  padding-bottom: 0.4em;
  background-color: var(--sklearn-color-background);
}

#sk-container-id-2 div.sk-container {
  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`
     but bootstrap.min.css set `[hidden] { display: none !important; }`
     so we also need the `!important` here to be able to override the
     default hidden behavior on the sphinx rendered scikit-learn.org.
     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */
  display: inline-block !important;
  position: relative;
}

#sk-container-id-2 div.sk-text-repr-fallback {
  display: none;
}

div.sk-parallel-item,
div.sk-serial,
div.sk-item {
  /* draw centered vertical line to link estimators */
  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));
  background-size: 2px 100%;
  background-repeat: no-repeat;
  background-position: center center;
}

/* Parallel-specific style estimator block */

#sk-container-id-2 div.sk-parallel-item::after {
  content: "";
  width: 100%;
  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);
  flex-grow: 1;
}

#sk-container-id-2 div.sk-parallel {
  display: flex;
  align-items: stretch;
  justify-content: center;
  background-color: var(--sklearn-color-background);
  position: relative;
}

#sk-container-id-2 div.sk-parallel-item {
  display: flex;
  flex-direction: column;
}

#sk-container-id-2 div.sk-parallel-item:first-child::after {
  align-self: flex-end;
  width: 50%;
}

#sk-container-id-2 div.sk-parallel-item:last-child::after {
  align-self: flex-start;
  width: 50%;
}

#sk-container-id-2 div.sk-parallel-item:only-child::after {
  width: 0;
}

/* Serial-specific style estimator block */

#sk-container-id-2 div.sk-serial {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: var(--sklearn-color-background);
  padding-right: 1em;
  padding-left: 1em;
}


/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is
clickable and can be expanded/collapsed.
- Pipeline and ColumnTransformer use this feature and define the default style
- Estimators will overwrite some part of the style using the `sk-estimator` class
*/

/* Pipeline and ColumnTransformer style (default) */

#sk-container-id-2 div.sk-toggleable {
  /* Default theme specific background. It is overwritten whether we have a
  specific estimator or a Pipeline/ColumnTransformer */
  background-color: var(--sklearn-color-background);
}

/* Toggleable label */
#sk-container-id-2 label.sk-toggleable__label {
  cursor: pointer;
  display: block;
  width: 100%;
  margin-bottom: 0;
  padding: 0.5em;
  box-sizing: border-box;
  text-align: center;
}

#sk-container-id-2 label.sk-toggleable__label-arrow:before {
  /* Arrow on the left of the label */
  content: "▸";
  float: left;
  margin-right: 0.25em;
  color: var(--sklearn-color-icon);
}

#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {
  color: var(--sklearn-color-text);
}

/* Toggleable content - dropdown */

#sk-container-id-2 div.sk-toggleable__content {
  max-height: 0;
  max-width: 0;
  overflow: hidden;
  text-align: left;
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-2 div.sk-toggleable__content.fitted {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

#sk-container-id-2 div.sk-toggleable__content pre {
  margin: 0.2em;
  border-radius: 0.25em;
  color: var(--sklearn-color-text);
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-2 div.sk-toggleable__content.fitted pre {
  /* unfitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {
  /* Expand drop-down */
  max-height: 200px;
  max-width: 100%;
  overflow: auto;
}

#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {
  content: "▾";
}

/* Pipeline/ColumnTransformer-specific style */

#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-2 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Estimator-specific style */

/* Colorize estimator box */
#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-2 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-2);
}

#sk-container-id-2 div.sk-label label.sk-toggleable__label,
#sk-container-id-2 div.sk-label label {
  /* The background is the default theme color */
  color: var(--sklearn-color-text-on-default-background);
}

/* On hover, darken the color of the background */
#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-unfitted-level-2);
}

/* Label box, darken color on hover, fitted */
#sk-container-id-2 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Estimator label */

#sk-container-id-2 div.sk-label label {
  font-family: monospace;
  font-weight: bold;
  display: inline-block;
  line-height: 1.2em;
}

#sk-container-id-2 div.sk-label-container {
  text-align: center;
}

/* Estimator-specific */
#sk-container-id-2 div.sk-estimator {
  font-family: monospace;
  border: 1px dotted var(--sklearn-color-border-box);
  border-radius: 0.25em;
  box-sizing: border-box;
  margin-bottom: 0.5em;
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-2 div.sk-estimator.fitted {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

/* on hover */
#sk-container-id-2 div.sk-estimator:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-2 div.sk-estimator.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Specification for estimator info (e.g. "i" and "?") */

/* Common style for "i" and "?" */

.sk-estimator-doc-link,
a:link.sk-estimator-doc-link,
a:visited.sk-estimator-doc-link {
  float: right;
  font-size: smaller;
  line-height: 1em;
  font-family: monospace;
  background-color: var(--sklearn-color-background);
  border-radius: 1em;
  height: 1em;
  width: 1em;
  text-decoration: none !important;
  margin-left: 1ex;
  /* unfitted */
  border: var(--sklearn-color-unfitted-level-1) 1pt solid;
  color: var(--sklearn-color-unfitted-level-1);
}

.sk-estimator-doc-link.fitted,
a:link.sk-estimator-doc-link.fitted,
a:visited.sk-estimator-doc-link.fitted {
  /* fitted */
  border: var(--sklearn-color-fitted-level-1) 1pt solid;
  color: var(--sklearn-color-fitted-level-1);
}

/* On hover */
div.sk-estimator:hover .sk-estimator-doc-link:hover,
.sk-estimator-doc-link:hover,
div.sk-label-container:hover .sk-estimator-doc-link:hover,
.sk-estimator-doc-link:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,
.sk-estimator-doc-link.fitted:hover,
div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,
.sk-estimator-doc-link.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

/* Span, style for the box shown on hovering the info icon */
.sk-estimator-doc-link span {
  display: none;
  z-index: 9999;
  position: relative;
  font-weight: normal;
  right: .2ex;
  padding: .5ex;
  margin: .5ex;
  width: min-content;
  min-width: 20ex;
  max-width: 50ex;
  color: var(--sklearn-color-text);
  box-shadow: 2pt 2pt 4pt #999;
  /* unfitted */
  background: var(--sklearn-color-unfitted-level-0);
  border: .5pt solid var(--sklearn-color-unfitted-level-3);
}

.sk-estimator-doc-link.fitted span {
  /* fitted */
  background: var(--sklearn-color-fitted-level-0);
  border: var(--sklearn-color-fitted-level-3);
}

.sk-estimator-doc-link:hover span {
  display: block;
}

/* "?"-specific style due to the `<a>` HTML tag */

#sk-container-id-2 a.estimator_doc_link {
  float: right;
  font-size: 1rem;
  line-height: 1em;
  font-family: monospace;
  background-color: var(--sklearn-color-background);
  border-radius: 1rem;
  height: 1rem;
  width: 1rem;
  text-decoration: none;
  /* unfitted */
  color: var(--sklearn-color-unfitted-level-1);
  border: var(--sklearn-color-unfitted-level-1) 1pt solid;
}

#sk-container-id-2 a.estimator_doc_link.fitted {
  /* fitted */
  border: var(--sklearn-color-fitted-level-1) 1pt solid;
  color: var(--sklearn-color-fitted-level-1);
}

/* On hover */
#sk-container-id-2 a.estimator_doc_link:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

#sk-container-id-2 a.estimator_doc_link.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-3);
}
</a></style><div id="sk-container-id-2" class="sk-top-container"><div class="sk-text-repr-fallback"><pre>GradientBoostingRegressor()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br>On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class="sk-container" hidden=""><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-2" type="checkbox" checked=""><label for="sk-estimator-id-2" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;&nbsp;GradientBoostingRegressor<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.4/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html">?<span>Documentation for GradientBoostingRegressor</span></a><span class="sk-estimator-doc-link fitted">i<span>Fitted</span></span></label><div class="sk-toggleable__content fitted"><pre>GradientBoostingRegressor()</pre></div> </div></div></div></div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>y_test_pred <span class="op">=</span> reg.predict(X_test)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>np.corrcoef(y_test, y_test_pred)[<span class="dv">0</span>][<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0.7091798896983967</code></pre>
</div>
</div>
<p>Not terrible. Let’s see if we can do better after hyperparameters optimization. For this, we can use the hyperopt package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> hyperopt <span class="im">import</span> fmin, hp, tpe, STATUS_OK</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> objective(n_estimators):</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Define algorithm</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    reg <span class="op">=</span>  GradientBoostingRegressor(n_estimators<span class="op">=</span>n_estimators)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Fit model</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># For demo purposes, we only use the first 100k samples</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    reg.fit(X_train.head(<span class="dv">100000</span>), y_train.head(<span class="dv">100000</span>))</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Uncomment this step to train the model on the full dataset:</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># reg.fit(X_train, y_train)</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Test model</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    y_test_pred <span class="op">=</span> reg.predict(X_test)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    correlation <span class="op">=</span> np.corrcoef(y_test, y_test_pred)[<span class="dv">0</span>][<span class="dv">1</span>]</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {<span class="st">'loss'</span>: <span class="op">-</span>correlation, <span class="st">'status'</span>: STATUS_OK}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>best_params <span class="op">=</span> fmin(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  fn<span class="op">=</span>objective,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  space<span class="op">=</span><span class="dv">10</span> <span class="op">+</span> hp.randint(<span class="st">'n_estimators'</span>, <span class="dv">980</span>),</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  algo<span class="op">=</span>tpe.suggest,</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  max_evals<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>100%|██████████| 10/10 [37:15&lt;00:00, 223.58s/trial, best loss: -0.75991576993057] </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>best_params</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'n_estimators': 966}</code></pre>
</div>
</div>
<p>Initially, the default value of 100 estimators was used. According to this hyperopt run, using 966 estimators results in a more performant model.</p>
<p>Now we can train the model again with this new hyperparameter value:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>reg <span class="op">=</span>  GradientBoostingRegressor(n_estimators<span class="op">=</span>best_params[<span class="st">"n_estimators"</span>])</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>X_train <span class="op">=</span> train_val_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>y_train <span class="op">=</span> train_val_encoded[<span class="st">"y_target"</span>]</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>X_test <span class="op">=</span> test_encoded.drop(columns<span class="op">=</span>[<span class="st">"spectrum_id"</span>, <span class="st">"b_target"</span>,  <span class="st">"y_target"</span>])</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>y_test <span class="op">=</span> test_encoded[<span class="st">"y_target"</span>]</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>reg.fit(X_train.head(<span class="dv">100000</span>), y_train.head(<span class="dv">100000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>#sk-container-id-3 {
  /* Definition of color scheme common for light and dark mode */
  --sklearn-color-text: black;
  --sklearn-color-line: gray;
  /* Definition of color scheme for unfitted estimators */
  --sklearn-color-unfitted-level-0: #fff5e6;
  --sklearn-color-unfitted-level-1: #f6e4d2;
  --sklearn-color-unfitted-level-2: #ffe0b3;
  --sklearn-color-unfitted-level-3: chocolate;
  /* Definition of color scheme for fitted estimators */
  --sklearn-color-fitted-level-0: #f0f8ff;
  --sklearn-color-fitted-level-1: #d4ebff;
  --sklearn-color-fitted-level-2: #b3dbfd;
  --sklearn-color-fitted-level-3: cornflowerblue;

  /* Specific color for light theme */
  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));
  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));
  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));
  --sklearn-color-icon: #696969;

  @media (prefers-color-scheme: dark) {
    /* Redefinition of color scheme for dark theme */
    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));
    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));
    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));
    --sklearn-color-icon: #878787;
  }
}

#sk-container-id-3 {
  color: var(--sklearn-color-text);
}

#sk-container-id-3 pre {
  padding: 0;
}

#sk-container-id-3 input.sk-hidden--visually {
  border: 0;
  clip: rect(1px 1px 1px 1px);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

#sk-container-id-3 div.sk-dashed-wrapped {
  border: 1px dashed var(--sklearn-color-line);
  margin: 0 0.4em 0.5em 0.4em;
  box-sizing: border-box;
  padding-bottom: 0.4em;
  background-color: var(--sklearn-color-background);
}

#sk-container-id-3 div.sk-container {
  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`
     but bootstrap.min.css set `[hidden] { display: none !important; }`
     so we also need the `!important` here to be able to override the
     default hidden behavior on the sphinx rendered scikit-learn.org.
     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */
  display: inline-block !important;
  position: relative;
}

#sk-container-id-3 div.sk-text-repr-fallback {
  display: none;
}

div.sk-parallel-item,
div.sk-serial,
div.sk-item {
  /* draw centered vertical line to link estimators */
  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));
  background-size: 2px 100%;
  background-repeat: no-repeat;
  background-position: center center;
}

/* Parallel-specific style estimator block */

#sk-container-id-3 div.sk-parallel-item::after {
  content: "";
  width: 100%;
  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);
  flex-grow: 1;
}

#sk-container-id-3 div.sk-parallel {
  display: flex;
  align-items: stretch;
  justify-content: center;
  background-color: var(--sklearn-color-background);
  position: relative;
}

#sk-container-id-3 div.sk-parallel-item {
  display: flex;
  flex-direction: column;
}

#sk-container-id-3 div.sk-parallel-item:first-child::after {
  align-self: flex-end;
  width: 50%;
}

#sk-container-id-3 div.sk-parallel-item:last-child::after {
  align-self: flex-start;
  width: 50%;
}

#sk-container-id-3 div.sk-parallel-item:only-child::after {
  width: 0;
}

/* Serial-specific style estimator block */

#sk-container-id-3 div.sk-serial {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: var(--sklearn-color-background);
  padding-right: 1em;
  padding-left: 1em;
}


/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is
clickable and can be expanded/collapsed.
- Pipeline and ColumnTransformer use this feature and define the default style
- Estimators will overwrite some part of the style using the `sk-estimator` class
*/

/* Pipeline and ColumnTransformer style (default) */

#sk-container-id-3 div.sk-toggleable {
  /* Default theme specific background. It is overwritten whether we have a
  specific estimator or a Pipeline/ColumnTransformer */
  background-color: var(--sklearn-color-background);
}

/* Toggleable label */
#sk-container-id-3 label.sk-toggleable__label {
  cursor: pointer;
  display: block;
  width: 100%;
  margin-bottom: 0;
  padding: 0.5em;
  box-sizing: border-box;
  text-align: center;
}

#sk-container-id-3 label.sk-toggleable__label-arrow:before {
  /* Arrow on the left of the label */
  content: "▸";
  float: left;
  margin-right: 0.25em;
  color: var(--sklearn-color-icon);
}

#sk-container-id-3 label.sk-toggleable__label-arrow:hover:before {
  color: var(--sklearn-color-text);
}

/* Toggleable content - dropdown */

#sk-container-id-3 div.sk-toggleable__content {
  max-height: 0;
  max-width: 0;
  overflow: hidden;
  text-align: left;
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-3 div.sk-toggleable__content.fitted {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

#sk-container-id-3 div.sk-toggleable__content pre {
  margin: 0.2em;
  border-radius: 0.25em;
  color: var(--sklearn-color-text);
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-3 div.sk-toggleable__content.fitted pre {
  /* unfitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

#sk-container-id-3 input.sk-toggleable__control:checked~div.sk-toggleable__content {
  /* Expand drop-down */
  max-height: 200px;
  max-width: 100%;
  overflow: auto;
}

#sk-container-id-3 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {
  content: "▾";
}

/* Pipeline/ColumnTransformer-specific style */

#sk-container-id-3 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-3 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Estimator-specific style */

/* Colorize estimator box */
#sk-container-id-3 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-3 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-2);
}

#sk-container-id-3 div.sk-label label.sk-toggleable__label,
#sk-container-id-3 div.sk-label label {
  /* The background is the default theme color */
  color: var(--sklearn-color-text-on-default-background);
}

/* On hover, darken the color of the background */
#sk-container-id-3 div.sk-label:hover label.sk-toggleable__label {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-unfitted-level-2);
}

/* Label box, darken color on hover, fitted */
#sk-container-id-3 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {
  color: var(--sklearn-color-text);
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Estimator label */

#sk-container-id-3 div.sk-label label {
  font-family: monospace;
  font-weight: bold;
  display: inline-block;
  line-height: 1.2em;
}

#sk-container-id-3 div.sk-label-container {
  text-align: center;
}

/* Estimator-specific */
#sk-container-id-3 div.sk-estimator {
  font-family: monospace;
  border: 1px dotted var(--sklearn-color-border-box);
  border-radius: 0.25em;
  box-sizing: border-box;
  margin-bottom: 0.5em;
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-0);
}

#sk-container-id-3 div.sk-estimator.fitted {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-0);
}

/* on hover */
#sk-container-id-3 div.sk-estimator:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-2);
}

#sk-container-id-3 div.sk-estimator.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-2);
}

/* Specification for estimator info (e.g. "i" and "?") */

/* Common style for "i" and "?" */

.sk-estimator-doc-link,
a:link.sk-estimator-doc-link,
a:visited.sk-estimator-doc-link {
  float: right;
  font-size: smaller;
  line-height: 1em;
  font-family: monospace;
  background-color: var(--sklearn-color-background);
  border-radius: 1em;
  height: 1em;
  width: 1em;
  text-decoration: none !important;
  margin-left: 1ex;
  /* unfitted */
  border: var(--sklearn-color-unfitted-level-1) 1pt solid;
  color: var(--sklearn-color-unfitted-level-1);
}

.sk-estimator-doc-link.fitted,
a:link.sk-estimator-doc-link.fitted,
a:visited.sk-estimator-doc-link.fitted {
  /* fitted */
  border: var(--sklearn-color-fitted-level-1) 1pt solid;
  color: var(--sklearn-color-fitted-level-1);
}

/* On hover */
div.sk-estimator:hover .sk-estimator-doc-link:hover,
.sk-estimator-doc-link:hover,
div.sk-label-container:hover .sk-estimator-doc-link:hover,
.sk-estimator-doc-link:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,
.sk-estimator-doc-link.fitted:hover,
div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,
.sk-estimator-doc-link.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

/* Span, style for the box shown on hovering the info icon */
.sk-estimator-doc-link span {
  display: none;
  z-index: 9999;
  position: relative;
  font-weight: normal;
  right: .2ex;
  padding: .5ex;
  margin: .5ex;
  width: min-content;
  min-width: 20ex;
  max-width: 50ex;
  color: var(--sklearn-color-text);
  box-shadow: 2pt 2pt 4pt #999;
  /* unfitted */
  background: var(--sklearn-color-unfitted-level-0);
  border: .5pt solid var(--sklearn-color-unfitted-level-3);
}

.sk-estimator-doc-link.fitted span {
  /* fitted */
  background: var(--sklearn-color-fitted-level-0);
  border: var(--sklearn-color-fitted-level-3);
}

.sk-estimator-doc-link:hover span {
  display: block;
}

/* "?"-specific style due to the `<a>` HTML tag */

#sk-container-id-3 a.estimator_doc_link {
  float: right;
  font-size: 1rem;
  line-height: 1em;
  font-family: monospace;
  background-color: var(--sklearn-color-background);
  border-radius: 1rem;
  height: 1rem;
  width: 1rem;
  text-decoration: none;
  /* unfitted */
  color: var(--sklearn-color-unfitted-level-1);
  border: var(--sklearn-color-unfitted-level-1) 1pt solid;
}

#sk-container-id-3 a.estimator_doc_link.fitted {
  /* fitted */
  border: var(--sklearn-color-fitted-level-1) 1pt solid;
  color: var(--sklearn-color-fitted-level-1);
}

/* On hover */
#sk-container-id-3 a.estimator_doc_link:hover {
  /* unfitted */
  background-color: var(--sklearn-color-unfitted-level-3);
  color: var(--sklearn-color-background);
  text-decoration: none;
}

#sk-container-id-3 a.estimator_doc_link.fitted:hover {
  /* fitted */
  background-color: var(--sklearn-color-fitted-level-3);
}
</a></style><div id="sk-container-id-3" class="sk-top-container"><div class="sk-text-repr-fallback"><pre>GradientBoostingRegressor(n_estimators=966)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br>On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class="sk-container" hidden=""><div class="sk-item"><div class="sk-estimator fitted sk-toggleable"><input class="sk-toggleable__control sk-hidden--visually" id="sk-estimator-id-3" type="checkbox" checked=""><label for="sk-estimator-id-3" class="sk-toggleable__label fitted sk-toggleable__label-arrow fitted">&nbsp;&nbsp;GradientBoostingRegressor<a class="sk-estimator-doc-link fitted" rel="noreferrer" target="_blank" href="https://scikit-learn.org/1.4/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html">?<span>Documentation for GradientBoostingRegressor</span></a><span class="sk-estimator-doc-link fitted">i<span>Fitted</span></span></label><div class="sk-toggleable__content fitted"><pre>GradientBoostingRegressor(n_estimators=966)</pre></div> </div></div></div></div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>y_test_pred <span class="op">=</span> reg.predict(X_test)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>np.corrcoef(y_test, y_test_pred)[<span class="dv">0</span>][<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0.7593069387129985</code></pre>
</div>
</div>
<p>Much better already. To get a more accurate view of the model performance, we should calculate the correlation per spectrum, instead of across the full dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>prediction_df_y <span class="op">=</span> pd.DataFrame({</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"spectrum_id"</span>: test_encoded[<span class="st">"spectrum_id"</span>],</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"target_y"</span>: y_test,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"prediction_y"</span>: y_test_pred,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>prediction_df_y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">spectrum_id</th>
<th data-quarto-table-cell-role="th">target_y</th>
<th data-quarto-table-cell-role="th">prediction_y</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>9.0</td>
<td>0.000000</td>
<td>0.006154</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>9.0</td>
<td>0.000000</td>
<td>0.008823</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>9.0</td>
<td>0.000000</td>
<td>0.010059</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>9.0</td>
<td>0.000000</td>
<td>0.011024</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>9.0</td>
<td>0.000000</td>
<td>0.019898</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">367683</td>
<td>398369.0</td>
<td>0.000000</td>
<td>0.262313</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">367684</td>
<td>398369.0</td>
<td>0.224074</td>
<td>0.268292</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">367685</td>
<td>398369.0</td>
<td>0.283664</td>
<td>0.208505</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">367686</td>
<td>398369.0</td>
<td>0.185094</td>
<td>0.177186</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">367687</td>
<td>398369.0</td>
<td>0.192657</td>
<td>0.263046</td>
</tr>
</tbody>
</table>

<p>367688 rows × 3 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>corr_y <span class="op">=</span> prediction_df_y.groupby(<span class="st">"spectrum_id"</span>).corr().iloc[::<span class="dv">2</span>][<span class="st">'prediction_y'</span>]</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>corr_y.index <span class="op">=</span> corr_y.index.droplevel(<span class="dv">1</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>corr_y <span class="op">=</span> corr_y.reset_index().rename(columns<span class="op">=</span>{<span class="st">"prediction_y"</span>: <span class="st">"correlation"</span>})</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>corr_y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">spectrum_id</th>
<th data-quarto-table-cell-role="th">correlation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>9.0</td>
<td>0.868330</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>16.0</td>
<td>0.725240</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>39.0</td>
<td>0.101923</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>95.0</td>
<td>0.118734</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>140.0</td>
<td>0.785892</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">27031</td>
<td>398328.0</td>
<td>0.675631</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">27032</td>
<td>398341.0</td>
<td>0.066689</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">27033</td>
<td>398342.0</td>
<td>0.763430</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">27034</td>
<td>398368.0</td>
<td>0.684804</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">27035</td>
<td>398369.0</td>
<td>-0.352181</td>
</tr>
</tbody>
</table>

<p>27036 rows × 2 columns</p>
</div>
</div>
</div>
<p>Median correlation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>corr_y[<span class="st">"correlation"</span>].median()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0.7524246447174651</code></pre>
</div>
</div>
<p>Correlation distribution:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>sns.set_style(<span class="st">"whitegrid"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sns.catplot(</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>corr_y, x<span class="op">=</span><span class="st">"correlation"</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>    fliersize<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>    kind<span class="op">=</span><span class="st">"box"</span>, aspect<span class="op">=</span><span class="dv">4</span>, height<span class="op">=</span><span class="dv">2</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="nist-2-traditional-ml-gradient-boosting_files/figure-html/cell-29-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Not bad! With some more hyperparameter optimization (optimizing only the number of trees is a bit crude) a lot more performance gains could be made. Take a look at the <a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html">Scikit Learn documentation</a> to learn more about the various hyperparameters for the <code>GradientBoostingRegressor</code>. Alternatively, you could switch to the <a href="https://xgboost.readthedocs.io/en/stable/">XGBoost algorithm</a>, which is currently used by MS²PIP.</p>
<p>And of course, this model can only predict y-ion intensities. You can repeat the training and optimization steps to train a model for b-ion intensities.</p>
<p>Good luck!</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Friedman2002" class="csl-entry" role="listitem">
Friedman, Jerome H. 2002. <span>“Stochastic Gradient Boosting.”</span> <em>Computational Statistics and Data Analysis</em> 38 (4): 367–78. <a href="https://doi.org/10.1016/s0167-9473(01)00065-2">https://doi.org/10.1016/s0167-9473(01)00065-2</a>.
</div>
<div id="ref-Gabriels2019" class="csl-entry" role="listitem">
Gabriels, Ralf, Lennart Martens, and Sven Degroeve. 2019. <span>“Updated MS²PIP Web Server Delivers Fast and Accurate MS² Peak Intensity Prediction for Multiple Fragmentation Methods, Instruments and Labeling Techniques.”</span> <em>NUCLEIC ACIDS RESEARCH</em> 47 (W1): W295–99. <a href="https://doi.org/10.1093/nar/gkz299">https://doi.org/10.1093/nar/gkz299</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/www\.proteomicsml\.org");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center"><small>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons CC BY 4.0</a> license.</small><div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>